Add support non-tile-multiple shard width

nilaykp opened the pull request:
### Ticket
[https://github.com/tenstorrent/tt-metal/issues/10109]

### Problem description
With tile multiple shard width, all cores might not be utilised which can degrade performance.

### What's changed
Add non-tile multiple shard width support.

### Checklist
- [X] Post commit CI passes [Link 1](https://github.com/tenstorrent/tt-metal/actions/runs/12009587806) [Link 2](https://github.com/tenstorrent/tt-metal/actions/runs/12078676939) [Link 3](https://github.com/tenstorrent/tt-metal/actions/runs/12138304841) [Link4](https://github.com/tenstorrent/tt-metal/actions/runs/12158369992) [Link5](https://github.com/tenstorrent/tt-metal/actions/runs/12166455738) [Latest](https://github.com/tenstorrent/tt-metal/actions/runs/12173859013)
- [x] Nightly fast dispatch [Link1](https://github.com/tenstorrent/tt-metal/actions/runs/12138312816) [Latest](https://github.com/tenstorrent/tt-metal/actions/runs/12158395787) [Latest](https://github.com/tenstorrent/tt-metal/actions/runs/12173862937) same as [main](https://github.com/tenstorrent/tt-metal/actions/runs/12174127726)
- [X] Model regression CI testing passes (if applicable) [Link1](https://github.com/tenstorrent/tt-metal/actions/runs/12138316171) [Link2](https://github.com/tenstorrent/tt-metal/actions/runs/12158409596) [Latest](https://github.com/tenstorrent/tt-metal/actions/runs/12173864263)
- [X] Device performance regression CI testing passes (if applicable) [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12138348929) [Latest](https://github.com/tenstorrent/tt-metal/actions/runs/12173865705)
- [ ] New/Existing tests provide coverage for changes

mywoodstock commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -907,8 +930,7 @@ def test_sd_conv_wh(
     if filter_height > 1 and (input_channels > 1280 or (input_channels > 640 and input_height > 16)):
-        if enable_auto_formatting:
-            pytest.skip("Not running split SD conv with auto formatting")
+        pytest.skip("Not running split SD conv with auto formatting")
```
why remove the if condition?

mywoodstock commented on ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/untilize/untilize_with_halo_op_v2.cpp:
```diff
@@ -0,0 +1,382 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
```
This file location has changed.

nkpatel-tt replied:
Done.

mywoodstock commented on ttnn/cpp/ttnn/operations/conv2d/conv2d.cpp:
```diff
@@ -89,7 +89,7 @@ ParallelConfig determine_parallel_config(
                 block_shard_orientation == ShardOrientation::COL_MAJOR ? device_grid_size[1] : device_grid_size[0];
             uint32_t num_cores_channels = find_closest_common_largest_divisor(
-                conv_out_2d_matrix_width_ntiles, std::ceil((double)input_channels / (double)32), total_cores_for_channels);
+                    conv_out_2d_matrix_width_ntiles, is_out_tiled ? std::ceil((double)input_channels / (double)32) : input_channels, total_cores_for_channels);
```
shouldn't `conv_out_2d_matrix_width_ntiles` by multiplied by tile width for non-tiled case?

mywoodstock commented on ttnn/cpp/ttnn/operations/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks_sharded.cpp:
```diff
@@ -0,0 +1,418 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
```
How is this compute kernel different from the existing one?

nkpatel-tt replied:
Removed this and made modifications in existing one

mywoodstock commented on ttnn/cpp/ttnn/operations/conv2d/device/kernels/reader_conv_activations_2d_mcast_padded_with_halo_3x3_weights_v2.cpp:
```diff
@@ -46,6 +46,7 @@ void kernel_main() {
     constexpr uint32_t window_inner = get_compile_time_arg_val(7);
     constexpr uint32_t act_block_h_datums = get_compile_time_arg_val(8);
     constexpr uint32_t weight_size_w = get_compile_time_arg_val(10);
+    constexpr uint32_t act_block_w_extra_align_bytes = get_compile_time_arg_val(12);
```
is this to align the sticks to 16B?

nkpatel-tt replied:
Yes, It is to align boundry to 32 bytes of single line(input_channels X 9) so that next pixel starts in next row of (n*h*w).

mywoodstock commented on ttnn/cpp/ttnn/operations/conv2d/device/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp:
```diff
@@ -1078,11 +1094,11 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_sharded_v2_impl(
         }
         else if (is_conv1d and is_depthwise_conv) {
-            // 1D Depthwise Conv
+            // height sharded Conv
```
this comment should remain same

mywoodstock commented on ttnn/cpp/ttnn/operations/conv2d/device/optimized_conv_op.hpp:
```diff
@@ -33,6 +33,22 @@ struct OptimizedConvParallelizationConfig {
     }
 };
 
+struct OptimizedConvParallelizationConfigNew {
```
can we use the existing config and update the values where used in tiled case?

nilaykp committed:
halo changes

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

tensor util changes

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

conv2d changes

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Normal test cases working with updated conv block config.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

kernel changes

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Resolve hang

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Clean up debug statements.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Clean up debug statements and functions.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Solve WS test hang.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Update input offset calculations based on alignment.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Align input matrix M for 1x1 conv according to new changes.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Fix after rebase

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Resolve Comments.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Remove Debug statements and commented code.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Changes with rebase.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Undo some changes after rebase-clenup

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Make pipeline work. Clean up needed.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Update input offset calculations based on alignment. Resolves maxpool and model pipeline failures.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Enable variable to support non-tile multiple width.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Divide tiles among cores instead of total height

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Modify Condition

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Change after latest rebase.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Modify test cases to accomodate small number of cores.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

Resolve Yolo failure.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

Add support for mulit-device tensor for weight and bias tensors.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

Add comment for prepare weight and bias matrix and modify condition.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

Remove debug statements.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

Address review comments

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:
```diff
@@ -5,6 +5,8 @@
 #include <array>
 #include <cstdint>
 #include "conv2d_op.hpp"
+#include <ostream>
```
remove

nkpatel-tt replied:
Done.

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_sharded_program_factory.cpp:
```diff
@@ -177,14 +181,26 @@ std::tuple<CBHandle, CBHandle> create_CBs_for_sharded_input_v2(
-            CircularBufferConfig(num_writer_output_tiles * out_tile_size, {{out0_cb, out_df}})
-                .set_page_size(out0_cb, out_tile_size);
-        if (output.is_sharded()) {
+        bool need_unpad_after_untilize = output_shard_shape[1] * output_shard_shape[0] < num_writer_output_tiles * 1024;
```
use tile size instead of hard-coding 1024.

nkpatel-tt replied:
Done.

shwetankTT commented on ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:
```diff
@@ -207,13 +207,13 @@ OptimizedConvParallelizationConfig determine_conv_op_parallel_config_from_conv_o
     const auto& shard_shape = shard_spec.shape;
     TT_ASSERT(conv_output_mem_config.memory_layout == TensorMemoryLayout::WIDTH_SHARDED || shard_shape[0] % 32 == 0);
-    TT_ASSERT(shard_shape[1] % 32 == 0);
+    /*TT_ASSERT(shard_shape[1] % 32 == 0);*/
```
Please remove the commented code

nkpatel-tt replied:
Done

shwetankTT commented:
@nkpatel-tt Please rebase the code to main. There are some changes in compute kernel make sure you are merging them properly.

shwetankTT commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:
```diff
@@ -18,6 +19,7 @@
 #include "ttnn/operations/experimental/auto_format/auto_format.hpp"
 
 #include "ttnn/operations/sliding_window/sliding_window.hpp"
+#include "ttnn/tensor/tensor.hpp"
```
is this include needed?

nkpatel-tt replied:
Removed

mywoodstock commented:
Also, please add unit tests.

pavlejosipovic commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -119,7 +120,7 @@ void MAIN {
     constexpr uint32_t tilized_in0_cb_id                        = tt::CB::c_intermed1;
     //constexpr uint32_t untilize_mode_reblock_cb                 = tt::CB::c_intermed2;
-    constexpr uint32_t out_cb_id                                = tt::CB::c_out0;
+    /*constexpr uint32_t out_cb_id                                = tt::CB::c_out0;*/
```
remove?

nkpatel-tt replied:
Done.

pavlejosipovic commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/reader_conv_activations_2d_mcast_padded_with_halo_3x3_weights_v2.cpp:
```diff
@@ -158,7 +159,9 @@ void kernel_main() {
             uint32_t two_reader_indices = packed_reader_indices_ptr[reader_idx];
             #if DILATION_W == 1
             read_channels(l1_write_addr_act, act_l1_read_addr, two_reader_indices & 0xffff, conv_act_c_read_bytes, coalesced_read_bytes, stride_h_bytes);
+            l1_write_addr_act += act_block_w_extra_align_bytes;
```
can we avoid doing this add if align is zero? with constexpr

nkpatel-tt replied:
Done.

nkpatel-tt commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -2417,3 +2417,101 @@ def test_non_tile_multiple_height_conv_wh(
         output_layout=ttnn.ROW_MAJOR_LAYOUT,
     )
+
+
```
Skipping as of now since tensor copy for non tile multiple width has problem(https://github.com/tenstorrent/tt-metal/issues/14081)

nkpatel-tt commented:
> Also, please add unit tests.

Done.

mywoodstock commented:
@nkpatel-tt looks like #14081 has been resolved, so this should be unblocked now.

mywoodstock commented on ttnn/cpp/ttnn/operations/data_movement/untilize_with_halo_v2/device/kernels/dataflow/halo_gather.cpp:
```diff
@@ -70,15 +71,23 @@ void copy_sticks_async(
                 uint64_t src_addr = base_addr + src_offset;
-                noc_async_read(src_addr, dst_addr, size);
+                for(uint16_t k = 0; k < nsticks; k++) {
+                    noc_async_read(src_addr, dst_addr, stick_nbytes);
```
No longer coalescing? I think we should have both versions since coalesced version is much higher performing.

nkpatel-tt replied:
Done.

nkpatel-tt commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -2621,6 +2621,86 @@ def test_non_tile_multiple_height_conv_wh(
+)
+@pytest.mark.parametrize(
+    "activations_dtype",
+    [ttnn.bfloat16],
```
For activations bfloat8_t for row major is not supported. Added for weights.

dmakoviichuk-tt replied:
you can add unsupported cases too. To make sure it will throw exception.

nkpatel-tt commented on ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:
```diff
@@ -103,8 +104,15 @@ ParallelConfig determine_parallel_config(
         uint32_t start_divisor =
                 block_shard_orientation == ShardOrientation::COL_MAJOR ? compute_grid_size.x : compute_grid_size.y;
+        auto channels_per_core = std::ceil((float)input_channels / effective_tile_width);
+        if(is_non_tile_mul_width) {
```
Yes as of now. For height it this feature is not needed and will raise separate PR for width sharding.

nkpatel-tt commented on ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:
```diff
@@ -103,8 +104,15 @@ ParallelConfig determine_parallel_config(
                 block_shard_orientation == ShardOrientation::COL_MAJOR ? compute_grid_size.x : compute_grid_size.y;
+        auto channels_per_core = std::ceil((float)input_channels / effective_tile_width);
+        if(is_non_tile_mul_width) {
+            out_nhw_ntiles = tt::round_up(batch_size * output_height * output_width, tt::constants::TILE_HEIGHT) / tt::constants::TILE_HEIGHT;
```
deleted.

nkpatel-tt commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -2621,6 +2621,86 @@ def test_non_tile_multiple_height_conv_wh(
+        (1, 512, 512, 16, 16, 3, 3, 1, 1, 1, 1, False, None),
+        (1, 512, 512, 32, 32, 3, 3, 1, 1, 1, 1, False, None),
+        (1, 640, 640, 16, 16, 3, 3, 1, 1, 1, 1, False, None),
+        (1, 640, 640, 32, 32, 3, 3, 1, 1, 1, 1, False, None),
```
Done.

nkpatel-tt commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -2621,6 +2621,86 @@ def test_non_tile_multiple_height_conv_wh(
+        and activations_dtype == ttnn.bfloat16
+        and weights_dtype == ttnn.bfloat16
+        and enable_auto_formatting == False
+    ):
```
Removed.

nkpatel-tt commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -2621,6 +2621,86 @@ def test_non_tile_multiple_height_conv_wh(
+        use_1d_systolic_array,
+        config_override,
+        use_shallow_conv_variant=(input_channels == 16),
+        transpose_mcast=use_1d_systolic_array,  ## use RM (transpose_mcast=False) with 2D on WH
```
Removed

nkpatel-tt commented on ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:
```diff
@@ -103,8 +104,15 @@ ParallelConfig determine_parallel_config(
+        auto channels_per_core = std::ceil((float)input_channels / effective_tile_width);
+        if(is_non_tile_mul_width) {
+            out_nhw_ntiles = tt::round_up(batch_size * output_height * output_width, tt::constants::TILE_HEIGHT) / tt::constants::TILE_HEIGHT;
+            out_c_ntiles = tt::round_up(output_channels, effective_tile_width) / 1;
```
This is needed since non-tile multiple height does not support Block sharding and is_out_tiled variable is used to pass status of non-tile multiple height variable.

dmakoviichuk-tt commented on ttnn/cpp/ttnn/tensor/tensor_utils.cpp:
```diff
@@ -210,6 +210,229 @@ Tensor convert_conv_weight_tensor_to_tiled_layout(
+
+// Converts convolution weights to tilized 2d matrix layout for block sharded conv.
+// Returns a new tensor with layout=Tile
+Tensor convert_conv_weight_tensor_to_tiled_layout_block_sharded(
```
Any chance to do it on device?
If not please create c++ test for this function.

nkpatel-tt replied:
As of now we are doing it on host. 
There are not tests for any function in this file(they are tested indirectly by conv2d op). I shall create separate issue for same.

mywoodstock replied:
There is a separate tracking for moving all weight pre-processing to device: #14080

nkpatel-tt replied:
Added test cases, pls check.

nilaykp committed:
Fix after rebase

nilaykp committed:
Move in_non_tile_mul_width variable to main function.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Remove extra exposed variable for non tile multiple width.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Remove std namespace.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

TT-BrianLiu commented on ttnn/cpp/ttnn/tensor/tensor_utils.cpp:
```diff
@@ -213,6 +213,234 @@ Tensor convert_conv_weight_tensor_to_tiled_layout(
+                output_buffer[matrix_idx] = input_buffer[idx];
+            }
+        }
+        if constexpr (std::is_same<T, float>::value) {
+            if (output_dtype == DataType::BFLOAT8_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp8_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
+            if (output_dtype == DataType::BFLOAT4_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp4_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
```
This looks like the same code copied/pasted again and again in this file. Can you please refactor this?

nkpatel-tt replied:
Done.

TT-BrianLiu commented on ttnn/cpp/ttnn/tensor/tensor_utils.cpp:
```diff
@@ -213,6 +213,234 @@ Tensor convert_conv_weight_tensor_to_tiled_layout(
+                }
+            }
+        }
+        if constexpr (std::is_same<T, float>::value) {
+            if (output_dtype == DataType::BFLOAT8_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp8_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
+            if (output_dtype == DataType::BFLOAT4_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp4_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
```
This looks like the same code copied/pasted again and again in this file. Can you please refactor this?

nkpatel-tt replied:
Done.

TT-BrianLiu commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/writer_tiled_out_2d_mcast_sender_conv_weights_tiled_col_to_rm_blocks.cpp:
```diff
@@ -4,7 +4,7 @@
 #include "dataflow_api.h"
 
-#define ENABLE_DEBUG 0
+#define ENABLE_DEBUG 1
```
Please remove

nkpatel-tt replied:
Done.

TT-BrianLiu commented on ttnn/cpp/ttnn/tensor/tensor_utils.cpp:
```diff
@@ -213,6 +213,234 @@ Tensor convert_conv_weight_tensor_to_tiled_layout(
+            // height padding for non tile multiple block height
+            weight_matrix_rows = weight_block_height_padded * num_channel_shards;
+        }
+        tt::tt_metal::LegacyShape output_shape = {1, 1, weight_matrix_rows, weight_matrix_cols};
```
`ttnn::SimpleShape`

nkpatel-tt replied:
Done.

TT-BrianLiu reviewed (changes requested):
Please address comments. In general, there is so much duplicate code in `ttnn/cpp/ttnn/tensor/tensor_utils.cpp`. Please think of a way to commonize code.

nilaykp committed:
Add test cases for weight and bias tensor convert functions.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

nilaykp committed:
tensor utils changes.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Address comment and remove unnecessary variable.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

nkpatel-tt commented:
> Please address comments. In general, there is so much duplicate code in `ttnn/cpp/ttnn/tensor/tensor_utils.cpp`. Please think of a way to commonize code.

Sure will check it out.

nilaykp committed:
Align tensor utils functions to new changes.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

ayerofieiev-tt commented on tests/tt_eager/ops/test_tensor_utils.cpp:
must be moved to somewhere like tests/ttnn/unit_tests/gtests/

TT-BrianLiu reviewed (approved):
Will approve, but please address this issue next: [#15691 In general, there is so much duplicate code in `ttnn/cpp/ttnn/tensor/tensor_utils.cpp`. Please think of a way to commonize code.](https://github.com/tenstorrent/tt-metal/issues/15691)

ayerofieiev-tt commented:
Thank you, @TT-BrianLiu ! I agree, this is really important.

nilaykp committed:
Address review comments.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Merge branch 'main' into nkpatel/conv_op_non_tile_multiple_shard_widht

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>

nilaykp committed:
Resolve conflicts after rebase.

Signed-off-by: Nilaykumar Patel <nkpatel@tenstorrent.com>
